### PR TITLE
reserve bock in scanner

### PR DESCRIPTION
**Pull request by jacktengg**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

File: be/src/vec/exec/scan/scanner_context.cpp [hunk 1]
```diff
@@ -232,9 +232,21 @@ vectorized::BlockUPtr ScannerContext::get_free_block(bool force) {
         // The caller of get_free_block will increase the memory usage
         update_peak_memory_usage(-block->allocated_bytes());
     } else if (_block_memory_usage < _max_bytes_in_queue || force) {
-        _newly_create_free_blocks_num->update(1);
-        block = vectorized::Block::create_unique(_output_tuple_desc->slots(), 0,
-                                                 true /*ignore invalid slots*/);
+        Status status;
+        if (!force && _state->enable_reserve_memory()) {
+            status = thread_context()->try_reserve_memory(block_avg_bytes);
+            if (!status.ok()) {
```
Comment (yiguolei) [inline be/src/vec/exec/scan/scanner_context.cpp:238]: reserve 之后，应该需要release reserverd
应该使用defer 那些macros

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 2]
```diff
@@ -2132,6 +2133,14 @@ public void setIgnoreShapePlanNodes(String ignoreShapePlanNodes) {
     @VariableMgr.VarAttr(name = MAX_FETCH_REMOTE_TABLET_COUNT, fuzzy = true)
     public int maxFetchRemoteTabletCount = 512;
 
+    @VariableMgr.VarAttr(
+            name = ENABLE_RESERVE_MEMORY,
+            description = {"控制是否启用分配内存前先reverve memory的功能。默认为 false。",
+                    "Controls whether to enable reserve memory before allocating memory. "
+                            + "The default value is false."},
+            needForward = true, fuzzy = true)
+    public boolean enableReserveMemory = false;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:2142]: 默认值改为true

File: be/src/vec/spill/spill_stream.cpp [hunk 3]
```diff
@@ -116,7 +116,9 @@ Status SpillStream::prepare() {
 }
 
 SpillReaderUPtr SpillStream::create_separate_reader() const {
-    return std::make_unique<SpillReader>(stream_id_, writer_->get_file_path());
+    return std::make_unique<SpillReader>(
+            state_->get_query_ctx()->get_mem_tracker()->get_query_statistics(), stream_id_,
```
Comment (yiguolei) [inline be/src/vec/spill/spill_stream.cpp:120]: 为啥改这个？

Comment (jacktengg) [reply be/src/vec/spill/spill_stream.cpp:120]: compile error

File: gensrc/thrift/PaloInternalService.thrift [hunk 4]
```diff
@@ -356,6 +359,11 @@ struct TQueryOptions {
   139: optional i32 query_slot_count = 0;
 
   140: optional bool enable_auto_create_when_overwrite = false;
+
+  141: optional bool enable_query_spill = false
```
Comment (yiguolei) [inline gensrc/thrift/PaloInternalService.thrift:363]: rename to enable_spill

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 5]
```diff
@@ -3863,9 +3839,8 @@ public TQueryOptions toThrift() {
         tResult.setParallelScanMinRowsPerScanner(parallelScanMinRowsPerScanner);
         tResult.setSkipBadTablet(skipBadTablet);
         tResult.setDisableFileCache(disableFileCache);
-        tResult.setEnableJoinSpill(enableJoinSpill);
-        tResult.setEnableSortSpill(enableSortSpill);
-        tResult.setEnableAggSpill(enableAggSpill);
+        tResult.setEnableReserveMemory(enableReserveMemory);
+        tResult.setEnableQuerySpill(enableQuerySpill);
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:3843]: rename to enablespill

File: be/src/pipeline/exec/streaming_aggregation_operator.cpp [hunk 6]
```diff
@@ -1154,7 +1154,7 @@ Status StreamingAggOperatorX::init(const TPlanNode& tnode, RuntimeState* state)
         _aggregate_evaluators.push_back(evaluator);
     }
 
-    if (state->enable_agg_spill()) {
+    if (state->enable_query_spill()) {
```
Comment (yiguolei) [inline be/src/pipeline/exec/streaming_aggregation_operator.cpp:1157]: rename to enable_spill

**Pull request merged**